### DISCUSSION TITLE
Keyboard handler stops being called for brief periods while typing

### DISCUSSION THREAD

**Discussion by KyleCardoza** [rsta2/circle#516] [Q&A] [answered]
I'm sure it's something I'm doing or not doing, but as I tinker with Circle, I have noticed that as I type on the rpi 400's keyboard continuously, for brief, ~1-2 second periods, it stops responding, then starts up again. No debug information is printed on the serial console when this happens. I was wondering if anyone else had seen this happen, and what could be done about it?

**Comment by KyleCardoza** [comment, marked as answer]
Never mind, I figured it out -- it turns out even a DMA burst setting of "1" is sufficient to monopolize the bus enough to mess with USB keyboards -- so it's back to updating the framebuffer with memcpy(), I guess.

**Reply by rsta2** [reply]
Have you tried a DMA burst length of 0? That's a valid parameter.

**Reply by KyleCardoza** [reply]
Fascinating. It failed with burst length 0 as well, until I randomly tried initializing the CDMAChannel instance with DMA_CHANNEL_NORMAL instead of DMA_CHANNEL_EXTENDED. With DMA_CHANNEL_NORMAL it works fine at burst 0, 1, and 2. I haven't tested higher with DMA_CHANNEL_NORMAL, since you said it usually causes problems.

So what, exactly, is the distinction between DMA_CHANNEL_NORMAL and DMA_CHANNEL_EXTENDED that might be the cause?

**Reply by rsta2** [reply]
`DMA_CHANNEL_NORMAL` uses the older 32-bit DMA engine, while `DMA_CHANNEL_EXTENDED` uses the 64-bit DMA4 engine, which was introduced with the RPi 4. Please read section 4.6. in [this document](https://datasheets.raspberrypi.com/bcm2711/bcm2711-peripherals.pdf) for details on the DMA4 engine.
